### PR TITLE
Leaking result cache metadata

### PR DESCRIPTION
**Pull request by pmelab**

### CODE REVIEW

File: src/Plugin/GraphQL/Fields/FieldPluginBase.php [hunk 1]
```diff
@@ -152,17 +152,18 @@ protected function isLanguageAwareField() {
    * {@inheritdoc}
    */
   protected function resolveDeferred(callable $callback, $value, array $args, ResolveContext $context, ResolveInfo $info) {
-    $result = $this->getRenderer()->executeInRenderContext(new RenderContext(), function () use ($callback, $value, $args, $context, $info) {
-      return $callback($value, $args, $context, $info);
-    });
+    $renderContext = new RenderContext();
+    $result = $callback($value, $args, $context, $info);
```
Comment (fubhy) [inline src/Plugin/GraphQL/Fields/FieldPluginBase.php]: Moving that outside of the render context kinda defeats the whole point, no?

**Review by fubhy — CHANGES_REQUESTED** [review]

File: src/Plugin/GraphQL/Fields/FieldPluginBase.php (hunk 1 above)
Comment (pmelab) [reply src/Plugin/GraphQL/Fields/FieldPluginBase.php]: The result is either a callable or a generator. And the generator is evaluated further down in `iterator_to_array`. Thats what I discovered when writing the tests.

Comment (pmelab) [reply src/Plugin/GraphQL/Fields/FieldPluginBase.php]: So thats what has to be wrapped in the render context.

Comment (fubhy) [reply src/Plugin/GraphQL/Fields/FieldPluginBase.php]: In that case, let's move the  line with new RenderContext() down there too.

Comment (pmelab) [reply src/Plugin/GraphQL/Fields/FieldPluginBase.php]: I'm okay with that 😃

**Comment by fubhy** [comment]
@pmelab I updated the PR. I think we need to wrap the whole thing.

File: src/Plugin/GraphQL/Fields/FieldPluginBase.php [hunk 2]
```diff
@@ -151,29 +151,34 @@ protected function isLanguageAwareField() {
...
-      return new Deferred(function () use ($result, $value, $args, $context, $info) {
-        return $this->resolveDeferred($result, $value, $args, $context, $info);
-      });
-    }
-
-    $result = iterator_to_array($result);
+  protected function resolveDeferred(callable $callback, $value, array $args, ResolveContext $context, ResolveInfo $info) {    
+    $renderContext = new RenderContext();
+    return $this->getRenderer()->executeInRenderContext($renderContext, function () use ($value, $args, $context, $info) {
+      $result = $callback($value, $args, $context, $info);
```
Comment (awm086) [inline src/Plugin/GraphQL/Fields/FieldPluginBase.php]: I think you need to add `$callback` and `$renderContext` to the use statement.

File: src/Plugin/GraphQL/Fields/FieldPluginBase.php [hunk 3]
```diff
@@ -151,29 +151,34 @@ protected function isLanguageAwareField() {
...
+        $dependencies = $this->getCacheDependencies($result, $value, $args, $context, $info);
+        foreach ($dependencies as $dependency) {
+          $context->addCacheableDependency($dependency);
+        }
       }
-    }
 
-    return $this->unwrapResult($result, $info);
+      return $this->unwrapResult($result, $info);
+    }
```
Comment (awm086) [inline src/Plugin/GraphQL/Fields/FieldPluginBase.php]: build fails for syntax error ... I think you're missing `);` here.

**Comment by fubhy** [comment]
I shouldn't try and edit this on my windows dual boot from the github text editor :D

**Review by awm086 — APPROVED** [review]

**Pull request merged**